crates/turborepo-repository/src/discovery.rs

```diff
@@ -216,69 +254,103 @@ impl<A: PackageDiscovery + Send, B: PackageDiscovery + Send> PackageDiscovery
             }
         }
     }
+
+    async fn discover_packages_blocking(&self) -> Result<DiscoveryResponse, Error> {
+        tracing::debug!("discovering packages using fallback strategy");
+
+        tracing::debug!("attempting primary strategy");
+        match tokio::time::timeout(self.timeout, self.primary.discover_packages_blocking()).await {
+            Ok(Ok(packages)) => Ok(packages),
+            Ok(Err(err1)) => {
+                tracing::debug!("primary strategy failed, attempting fallback strategy");
+                match self.fallback.discover_packages_blocking().await {
+                    Ok(packages) => Ok(packages),
+                    // if the backup is unavailable, return the original error
+                    Err(Error::Unavailable) => Err(err1),
+                    Err(err2) => Err(err2),
+                }
+            }
+            Err(_) => {
```

Comment:
Nit: I think the `Elapsed` error is worth specifying here, mostly as a guard against the code changing and forcing authors to handle any new errors they introduce, since the error message here is specific to a timeout.

Reply:
Good point, I have made the match total